pkg/transportrequest/gitutils.go

```diff
@@ -2,12 +2,73 @@ package transportrequest
 
 import (
 	"fmt"
+	gitUtils "github.com/SAP/jenkins-library/pkg/git"
+	"github.com/SAP/jenkins-library/pkg/log"
 	"github.com/SAP/jenkins-library/pkg/piperutils"
+	"github.com/go-git/go-git/v5"
 	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/pkg/errors"
+	"os"
 	"regexp"
 	"sort"
 )
 
+var logRange = gitUtils.LogRange
+var findLabelsInCommits = FindLabelsInCommits
+
+type iTransportRequestGitUtils interface {
+	PlainOpen(directory string) (*git.Repository, error)
+}
+type transportRequestGitUtils struct {
+}
+
+func (g *transportRequestGitUtils) PlainOpen(directory string) (*git.Repository, error) {
```

Comment:
What is the benefit of having another PlainOpen in the general git utils? Why not using the PlainOpen from the git utils directly?

Reply:
That's the way how the mocking works. Since TestFindIDInRange() needs to mock PlainOpen() it passes the struct as interface to findIDInRange().

Reply:
I made a mistake in my question above. Should be

What is the benefit of having another PlainOpen **here**?

I don't fully understand why we have the second `PlainOpen` here. Couldn't we provide a mocked git utils for the util in `pkg/git`. Maybe there is a good reaon for doing it like it is done, but I would like to understand ...

Reply:
You cants easily access the types of another test scripts. Go does not import the _test.go files. Instead, you'ld have to create something like a pkg/test/common.go where you put your common types. That is a decision, I cannot take, therefore I decided to implement a "redundant" approach.

Reply:
> Instead, you'ld have to create [...] That is a decision, I cannot take

We do that in other cases already, seach for `grep -r "+build" pkg`.

All this file are _in principle_ available - not only on the level of the tests. 

Of course test files needs to be separated from "payload" files. In other languages that separation is done by having dedicated folders for the test coding. Since that is not possible here this way we introduced the "annottion": `// +build !release` "annotation". In the official build all files having that annotation are excluded so that they are not part of the binary. See [here](https://github.com/SAP/jenkins-library/blob/master/Dockerfile#L18). In case some references something in one of these additional test files the official build will break.

Reply:
The files I found implement the mocking differenty.  Interfaces aren't used as far as I saw. Here, we have an interface, the real and the mock. the interface and the real have to be available to all, so we cant annotate them out. So we have to split into

```
interface/git.go
mock/git.go
```

The imports would look like

```
pkg/git/git.go -> imports interface/git.go
pkg/git/git_test.go -> imports interface/git.go and mock/git.go
pkg/transportrequest/gitutils.go -> imports interface/git.go 
pkg/transportrequest/gitutils_test.go -> imports interface/git.go and mock/git.go
```

Reply:
Yes, interface and real impl always available and the mock in either a `_test.go` file or in a file taged with `+build !release`.

For the mock we have a mock package: `pkg/mock`. And some colleagues prefer to have the mocks in some dedicated files beside the "payload" coding, like e.g. .here: `pkg/whitesource/utilsMock.go`, `pkg/config/mocks/vaultClient.go`.

> Interfaces aren't used as far as I saw

What do you mean with that? We use interfaces almost everywhere as far as I can see. E.g. for the file mock (`pkg/mock/fileUtils.go `) we have interfaces as we need it in the concrete use case, e.g. [here](https://github.com/SAP/jenkins-library/blob/master/cmd/cloudFoundryDeploy.go#L25). Maybe the (very) late binding causes some confusion here ...

Reply:
The sample, you point to, I dont understand. It doesnt use `pkg/mock/fileUtils.go` as you said. Instead, it uses `piperutils/fileUtils.go`. furthermore, it creates an own interface `type cfFileUtil interface`, although fileUtils provids such `type FileUtils interface`.

And the mocking layer is done by a global variable

`var fileUtils cfFileUtil = piperutils.Files{}`

Thats a different approach.

Reply:
`vault.go.` is a good sample. It uses the interface abstraction layer. But since it is the only consumer of the `vaultClient`, the interface is defined within `vault.go` and not put into a common file.

I mean, I can split it...

Reply:
> It doesnt use `pkg/mock/fileUtils.go`

It does, the mock is injected [here](https://github.com/SAP/jenkins-library/blob/master/cmd/cloudFoundryDeploy_test.go#L61).

> Instead, it uses `piperutils/fileUtils.go`

This is the "payload" instance we use in the free wild.

> it creates an own interface

That is - to my understanding - the idiomatic way in go with a very very late binding.

> although fileUtils provides such type `FileUtils` interface

Maybe it would be better to remove that interface since it seems to be more idiomatic when the use of a certain type defined the interface ...

> And the mocking layer is done by a global variabe

There are currently two approaches in the coding. Thisone with a global variable and anotherone where a interface is defined which collects **everything**. That interface is backed by a utils class for the real world and by anotherone for the tests. That utils gets handed-in into the runXXX method. That is part of the generated coding, an example is e.g. [here](https://github.com/SAP/jenkins-library/blob/master/cmd/hadolintExecute.go#L66). I personally dislike that approach to some extend, since it is not clear at all which functionality is contained in that utils class. But there are colleagues who **love** that approach. And: does it really make a difference (beside the ability to execute tests in parallel) if we use a variable or an additional parameter in a function signature?